Throw error message on submit in "Add Item" form when duplicate

dkimlim opened the pull request:
- useEffect to pull an array of all documents in the current token's collection, and set a shoppingList state with the array
- on handleSubmit, check if there's a duplicate in the shoppingList array state
- throw error message in div when there's duplicate
- normalize name in useEffect on submit (so data in firebase is not affected)

Stretch goals:
- show field that needs to be changed to user on error (ex. focus on item name input field)
- redirect to AddItem page on submit
- add better css for error message

jendevelops committed:
Add getCollection method, add shopping list collection local state

dkimlim committed:
change firebase call to useEffect and change default shoppingListCollection to an array of all items from colelction

dkimlim committed:
Checked for duplicate itemName in the handleSubmit

dkimlim committed:
created state for Error Message

dkimlim committed:
Added error message on submit if there's a duplicate item

dkimlim committed:
Removed uuid package from form submit data

dkimlim committed:
Updated object id to be the same as doc

dkimlim committed:
Normalize name when getting data in useEffect

dkimlim committed:
Verify normalized names on the submit action

AshlingH commented on src/components/AddItemForm.js:
```diff
@@ -37,11 +37,11 @@ const Form = ({token}) => {
                 console.log("Error getting document:", error);
             });
-    };
+    }, []);
```
would love some discussion on it on Sunday if possible, remind me again the purpose

jendevelops replied:
The purpose of the empty array is to have a "default" value for  useEffect. This triggers useEffect only after the first render, to prevent overloading requests to the Firebase API.

laurenmbeatty replied:
I think we should put in the dependency array: `[token, shoppingListCollection]`. This will run the useEffect if the shoppingListCollection updates or the token updates (i know the token probably won't really update, but my vs code eslint is yelling at me to include it in the array.  : )). Then...on line 49...see my comment.

jendevelops replied:
Thanks for that tip, Lauren. I now actually understand what the dependency array is used for. Completely makes sense now.

AshlingH commented:
Guys really excellent Pull request, would love to know more about the normalize names on Submit. Your code was very helpful in compiling our code! Well done

jonbascos reviewed (commented):
Looks good.  Just so I'm understanding, we are using the id given by Firestore instead of implementing uuid?  Is that what I'm seeing?  Otherwise, looks good as always.

jendevelops commented on src/components/AddItemForm.js:
```diff
@@ -1,28 +1,60 @@
+                lastPurchaseDate
+            };
+            tokenRef.add(data)
+            .then((docRef) => { tokenRef.doc(docRef.id).update({ id : docRef.id });  })
```
> Looks good. Just so I'm understanding, we are using the id given by Firestore instead of implementing uuid? Is that what I'm seeing? Otherwise, looks good as always.

That is correct. The add data request to Firebase returns a docRef response. If you take a look line 49, we then await on the docRef response and take the ID value of that response and assign it to our document, eliminating the unnecessary use of the uuid package.

laurenmbeatty replied:
```suggestion
            .then((docRef) => { tokenRef.doc(docRef.id).update({ id : docRef.id });  setShoppingListCollection(shoppingListCollection.concat(data))})
```
This will trigger the useEffect above so that you have the latest shopping list items from the db to compare.

Maybe at the beginning of the `handleSubmit` we could `setDuplicateError(false)` so that the error goes away when first submitting an item. It can be confusing because the error is always there once it's triggered.

jendevelops replied:
If the new data is joined with the shoppingListCollection to trigger useEffect, won't shoppingListCollection be overwritten with the Firebase call in useEffect?

laurenmbeatty replied:
I think so, yes. So I'm thinking of this situation:

1. List loads with apples, bananas, garlic
2. I successfully add garbage bags to the list via the add item form.
3. Right now if I try to enter garbage bags again, it wouldn't see it is a duplicate, because react state doesn't know that garbage bags is in the list (it only knows about apples, bananas and garlic). 
4. You can do it in a different way, but after adding an item to the list, you want to make sure your local state has that item in the list, too. Otherwise, you'll be able to add that item over and over again.

jendevelops replied:
That makes sense. I was thinking that it would redirect back to /AddItem on Submit, but we have a prevent default up there at the top of handleSubmit.

laurenmbeatty commented on src/components/AddItemForm.js:
```diff
@@ -1,28 +1,60 @@
 import '../css/AddItemForm.css';
-import { v4 as uuidv4 } from 'uuid';
+// import { Redirect } from 'react-router-dom';
 
```
If this isn't being used, we can just remove it.

laurenmbeatty commented on src/components/AddItemForm.js:
```diff
@@ -1,28 +1,60 @@
+                    let documentData = doc.data();
+                    let nameData = documentData.itemName
+                    nameData = nameData.toLowerCase().replace(/[.,\/#!$%\^&\*;:{}=\-_`~()]/g,"").trim().replace(/\s{2,}/g," ");
+
```
Did this come from somewhere?  How does this work if there is nothing in the db? Like if there is no nameData, it might error on `nameData.toLowerCase()`, I wonder.

laurenmbeatty replied:
After you successfully add an item, does this useEffect run again? We would want it to, so that your `shoppingListCollection` is updated.

jendevelops replied:
> Did this come from somewhere? How does this work if there is nothing in the db? Like if there is no nameData, it might error on `nameData.toLowerCase()`, I wonder.

That's a good point. We should put in a conditional to make sure the db isn't empty.

laurenmbeatty reviewed (commented):
This looks amazing you all. I had a couple of questions I wasn't sure about, and I haven't officially pulled it down to test it out yet.

jendevelops committed:
Add token and shoppingListCollection as trigger for useEffect

jendevelops committed:
Set duplicate error to false on handleSubmit, augment shoppingListCollection to change on add item

jendevelops committed:
Add conditional in use
Effect to check for empty collection and empty itemName

dkimlim committed:
Required fields for itemName and purchaseDate

laurenmbeatty commented on src/components/AddItemForm.js:
```diff
@@ -1,28 +1,64 @@
+                console.log("Error getting document:", error);
+            });
+    }, [token, shoppingListCollection]);
 
```
My super bad...we need to remove `shoppingListCollection` from this array, because it is overloading the db. When  the component loads it sets an initial state to `[]`, and then in this `useEffect`, it sets `shoppingListCollection` to whatever is in the db, and then it re-renders and sets `shoppingListCollection` to an empty array, etc. etc. etc. We still need to figure out a way to get the new item into the `shoppingListCollection`, though.  Sorry! @jbatara and @dkimlim . I will think on it if you all can, too.

laurenmbeatty replied:
I've got an idea that actually works and doesn't overload the db.  : ) Note, this is AN approach, and it seems to work with both an existing list and a brand new list, but you may come up with a better way, for sure.

1. Move the code inside the `useEffect` into a function like this:
`const getCurrentShoppingListItems = (userToken) => {
   const db = fb.firestore()
   // etc. etc. --- all of the code that's inside the useEffect
}`

2.Inside the `useEffect` call this function: `getCurrentShoppingListItems(userToken)` ---this will ensure that we get the db items on page load.
3.pass `userToken` into the `useEffect` dependency array.
4. Inside your `.then` of the `handleSubmit` call `getCurrentShoppingListItems(userToken)` something like this:
`.then((docRef) => { tokenRef.doc(docRef.id).update({ id : docRef.id });  getCurrentShoppingListItems(userToken)})`
It needs to be inside the `.then`. This ensures that the item for sure has been added to the db before we pull back the new list.
5. Test this out locally.  Never assume that your mentor/code reviewer is correct.  : )

Let me know if you want to zoom, @jbatara and @dkimlim

laurenmbeatty commented on src/components/AddItemForm.js:
```diff
@@ -1,28 +1,64 @@
+                    querySnapshot.forEach((doc) => {
+                        let documentData = doc.data();
+                        let nameData = documentData.itemName
+                        if(!nameData){
```
This should be 
```suggestion
                        if(nameData){
```
you want it to run this if there is `nameData`, right?

laurenmbeatty commented on src/components/AddItemForm.js:
```diff
@@ -34,7 +70,9 @@ const Form = ({token}) => {
-            onChange={e => setItemName(e.target.value)}/>
+            onChange={e => setItemName(e.target.value)}
+            required
+            />
```
Adding these `required` fields is awesome.

jendevelops committed:
Remove bang operator from nameData conditional

Fix incorrect logic to run normalization on itemNames that exist

Co-Authored-By: Lauren Beatty <laurenmbeatty@users.noreply.github.com>

dkimlim committed:
Fix the userEffect call to not have multiple calls

laurenmbeatty commented:
OK, this code totally works and meets the AC...amazing work, you two! Now we have to figure out why the build is failing. Hrrm. I don't see anything in the build logs that means anything to me, honestly. Can you try moving the whole function definition for `getCurrentShoppingListItems above the `useEffect` ? That way we can at least try re-triggering a build. 🥇 It is definitely not over-triggering the db anymore, so that's good!  : )

jendevelops committed:
Add addStatus and setAddStatus state and conditional render

Add local state addStatus for the status of adding a new item and
redirects back to the same page when item successfully added. Also
move getCurrentShoppingListItems above useEffect

joeylaguna commented:
Great work! The code ya'll have written is quite easy to read, so reviewing it was a breeze. Approved!

joeylaguna reviewed (approved):
🎉

jendevelops commented:
The only problem we're having now is the form is not clearing on redirect when an item is added to the Firebase collection. Maybe this is something we can discuss in Sunday meeting.

jendevelops commented:
Closes issue #7
